google/auth/metrics.py

```diff
@@ -0,0 +1,165 @@
+    )
+
+
+# x-goog-api-client header value for impersonated credentials access token request.
+# Example: "gl-python/3.7 auth/1.1 auth-request-type/at cred-type/imp"
+def token_request_access_token_impersonate():
+    return "{} {} {} {}".format(
+        python_version(),
+        auth_lib_version(),
+        REQUEST_TYPE_ACCESS_TOKEN,
+        CRED_TYPE_SA_IMPERSONATE,
+    )
+
+
+# x-goog-api-client header value for impersonated credentials ID token request.
+# Example: "gl-python/3.7 auth/1.1 auth-request-type/it cred-type/imp"
+def token_request_id_token_impersonate():
+    return "{} {} {} {}".format(
+        python_version(),
+        auth_lib_version(),
+        REQUEST_TYPE_ID_TOKEN,
+        CRED_TYPE_SA_IMPERSONATE,
+    )
+
+
+# x-goog-api-client header value for service account credentials access token
+# request (assertion flow).
+# Example: "gl-python/3.7 auth/1.1 auth-request-type/at cred-type/sa"
+def token_request_access_token_sa_assertion():
+    return "{} {} {} {}".format(
+        python_version(),
+        auth_lib_version(),
+        REQUEST_TYPE_ACCESS_TOKEN,
+        CRED_TYPE_SA_ASSERTION,
+    )
+
+
+# x-goog-api-client header value for service account credentials ID token
+# request (assertion flow).
+# Example: "gl-python/3.7 auth/1.1 auth-request-type/it cred-type/sa"
+def token_request_id_token_sa_assertion():
+    return "{} {} {} {}".format(
+        python_version(),
+        auth_lib_version(),
+        REQUEST_TYPE_ID_TOKEN,
+        CRED_TYPE_SA_ASSERTION,
+    )
+
+
+# x-goog-api-client header value for user credentials token request.
+# Example: "gl-python/3.7 auth/1.1 cred-type/u"
+def token_request_user():
+    return "{} {} {}".format(python_version(), auth_lib_version(), CRED_TYPE_USER)
+
+
+# Miscellenous metrics
+
+# x-goog-api-client header value for metadata server ping.
+# Example: "gl-python/3.7 auth/1.1 auth-request-type/mds"
+def mds_ping():
+    return "{} {} {}".format(
+        python_version(), auth_lib_version(), REQUEST_TYPE_MDS_PING
+    )
+
+
+# x-goog-api-client header value for reauth start endpoint calls.
+# Example: "gl-python/3.7 auth/1.1 auth-request-type/re-start"
+def reauth_start():
+    return "{} {} {}".format(
+        python_version(), auth_lib_version(), REQUEST_TYPE_REAUTH_START
+    )
+
+
+# x-goog-api-client header value for reauth continue endpoint calls.
+# Example: "gl-python/3.7 auth/1.1 cred-type/re-cont"
+def reauth_continue():
+    return "{} {} {}".format(
+        python_version(), auth_lib_version(), REQUEST_TYPE_REAUTH_CONTINUE
+    )
+
+
+def add_metric_header(headers, metric_header_value):
+    """Add x-goog-api-client header with the given value.
+
+    Args:
+        headers (Mapping[str, str]): The headers to which we will add the
+            metric header.
+        metric_header_value (Optional[str]): If value is None, do nothing;
+            if headers already has a x-goog-api-client header, append the value
```

Comment:
If the x-goog-api-client header is already present shouldn there be a check for the individual parts of the header? I assume if the header already exists the language version section would already be there for example

Reply:
For the token refresh requests, there is no x-goog-api-client header so we are fine.

For api requests where we add the token, client lib adds `x-goog-api-client` header. From auth lib side, we just append `cred-type/xx` to the header. 
- We don't need to send lang version because the header already has it
- We don't send auth version per discussion with concord team, because this will dramatically increase the size of concord table. For example, if currently we have a row for pubsub api call and the request_count field value is 100, it's likely these api calls use 5 different auth versions, 20 for each version, then the new table will need 5 rows, one row for each auth version).